sound/soc/sof/sof-ssp-audio.c

```diff
@@ -0,0 +1,229 @@
+
+	machine->mach_params.platform = dev_name(&pdev->dev);
+
+	sof_audio->machine = machine;
+
+	return 0;
+#endif
+}
+
+static int sof_audio_select_machine(struct platform_device *pdev,
+				    const struct sof_dev_desc *desc)
+{
+	struct sof_audio_dev *sof_audio = sof_mfd_get_client_data(&pdev->dev);
+	struct snd_soc_acpi_mach *mach;
+	int ret;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE)
+	/* force nocodec mode */
+	dev_warn(&pdev->dev, "Force to use nocodec mode\n");
+	mach = devm_kzalloc(&pdev->dev, sizeof(*mach), GFP_KERNEL);
+	if (!mach)
+		return -ENOMEM;
+
+	ret = sof_nocodec_setup(&pdev->dev, sof_audio, mach, desc);
+	if (ret < 0)
+		return ret;
+	mach->mach_params.platform = dev_name(&pdev->dev);
+#else
+	/* find machine */
+	mach = snd_soc_acpi_find_machine(desc->machines);
+	if (!mach) {
+		dev_warn(&pdev->dev, "warning: No matching ASoC machine driver found\n");
+	} else {
+		mach->mach_params.platform = dev_name(&pdev->dev);
+		sof_audio->tplg_filename = mach->sof_tplg_filename;
+	}
+#endif /* CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE */
+
+	sof_audio->machine = mach;
+	sof_audio->tplg_filename_prefix = desc->default_tplg_path;
+
+	/* check machine info */
+	ret = sof_machine_check(pdev, desc);
+	if (ret < 0)
+		dev_warn(&pdev->dev, "warning: failed to get machine info %d\n",
+			 ret);
+
+	return ret;
+}
+
+static int sof_audio_probe(struct platform_device *pdev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(pdev->dev.parent);
+	struct snd_sof_pdata *plat_data = sdev->pdata;
+	struct sof_mfd_client *audio_client = dev_get_platdata(&pdev->dev);
+	const struct sof_dev_desc *desc = plat_data->desc;
+	struct sof_audio_dev *sof_audio;
+	struct ipc_rx_client *audio_rx;
+	const char *drv_name;
+	const void *machine;
+	int num_drv;
+	int size;
+	int ret;
+
+	/* set IPC RX and TX reply callbacks */
+	audio_client->sof_client_rx_message = sof_audio_rx_message;
+	audio_client->sof_ipc_reply_cb = NULL;
+
+	/* create SOF audio device */
+	sof_audio = devm_kzalloc(&pdev->dev, sizeof(*sof_audio), GFP_KERNEL);
+	if (!sof_audio)
+		return -ENOMEM;
+
+	INIT_LIST_HEAD(&sof_audio->pcm_list);
+	INIT_LIST_HEAD(&sof_audio->kcontrol_list);
+	INIT_LIST_HEAD(&sof_audio->widget_list);
+	INIT_LIST_HEAD(&sof_audio->dai_list);
+	INIT_LIST_HEAD(&sof_audio->route_list);
+
+	sof_audio->audio_ops = desc->audio_ops;
+	sof_audio->platform = dev_name(&pdev->dev);
+
+	/* check for mandatory audio ops */
+	if (!sof_audio || !sof_audio->audio_ops->ipc_pcm_params)
+		return -EINVAL;
+
+	audio_client->client_data = sof_audio;
+
+	/* select machine driver */
+	ret = sof_audio_select_machine(pdev, desc);
+	if (ret < 0)
+		return ret;
+
+	/* set up platform component driver */
+	snd_sof_new_platform_drv(sof_audio);
+
+	/* number of dai drv to register */
+	if (!strcmp(sof_audio->machine->drv_name, "sof-nocodec"))
+		num_drv = sof_audio->audio_ops->num_ssp_drv +
+				sof_audio->audio_ops->num_dmic_drv;
+	else
+		num_drv = sof_audio->audio_ops->num_ssp_drv;
+
```

Comment:
This looks like a hack to work on Intel platforms.

Reply:
fixed.